prune.go

```diff
@@ -242,47 +277,86 @@ type CleanupStats struct {
 	ShouldRetry bool
 }
 
+func startReportingThread(stop <-chan struct{}, progressCh chan ProgressMessage) {
+	tStart := time.Now()
+	next := time.After(15 * time.Second)
+	go func() {
+		txnsRemoved := 0
+		docsCleaned := 0
+		for {
+			select {
+			case <-stop:
+				return
+			case msg := <-progressCh:
+				txnsRemoved += msg.TxnsRemoved
+				docsCleaned += msg.DocsCleaned
+			case <-next:
+				txnRate := 0.0
+				since := time.Since(tStart).Seconds()
+				if since > 0 {
+					txnRate = float64(txnsRemoved) / since
+				}
+				logger.Debugf("pruning has removed %d txns (%.0ftxn/s) cleaning %d docs ",
+					txnsRemoved, txnRate, docsCleaned)
+				next = time.After(15 * time.Second)
+			}
+		}
+	}()
+}
+
 // CleanAndPrune runs the cleanup steps, and then follows up with pruning all
 // of the transactions that are no longer referenced.
 func CleanAndPrune(args CleanAndPruneArgs) (CleanupStats, error) {
+	tStart := time.Now()
```

Comment:
The CR checklist suggests using utils.Clock instead of time.Now or time.After, though the reason may be lost in history.

Reply:
So, if you ever want to *test* something, you shouldn't depend on the wall clock. By using a Clock it gives you an interface to inject artificial time.
I don't think it particularly matters for 'how long did this operation take' which is just an informative message to the user. Though I probably should thread through a Clock and add a test around TxnBatchSleepTime. Thanks for the reminder.